cdap-api-common/src/main/java/io/cdap/cdap/api/data/format/StructuredRecord.java

```diff
@@ -381,6 +407,32 @@ public Builder setTimestamp(String fieldName, @Nullable ZonedDateTime zonedDateT
       }
     }
 
+    /**
+     * Sets the decimal value for provided {@link LogicalType#DECIMAL} field.
+     *
+     * @param fieldName name of the field to set
+     * @param decimal value for the field
+     * @return this builder
+     * @throws UnexpectedFormatException if the field is not in the schema, or the field is not nullable but a null
+     *                                   value is given or if the provided decimal is invalid
+     */
+    public Builder setDecimal(String fieldName, @Nullable BigDecimal decimal) {
+      validateAndGetLogicalTypeSchema(validateAndGetField(fieldName, decimal), EnumSet.of(LogicalType.DECIMAL));
+      if (decimal == null) {
+        fields.put(fieldName, null);
+        return this;
+      }
+
+      if (decimal.precision() <= 0) {
```

Comment:
This should be checking that the decimal scale and precision match those expected by the schema. The schema already checks that precision is positive, so this scenario will also be covered.

Please also add a unit test for this scenario.